src/jvm/main/org/jetbrains/kotlinx/lincheck/ObjectTraverser.kt

```diff
@@ -0,0 +1,146 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Lesser Public License for more details.
+ *
+ * You should have received a copy of the GNU General Lesser Public
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/lgpl-3.0.html>
+ */
+
+package org.jetbrains.kotlinx.lincheck
+
+import org.jetbrains.kotlinx.lincheck.LincheckClassLoader.REMAPPED_PACKAGE_CANONICAL_NAME
+import org.jetbrains.kotlinx.lincheck.strategy.managed.getObjectNumber
+import java.lang.reflect.Modifier
+import java.math.BigDecimal
+import java.math.BigInteger
+import java.util.*
+import java.util.concurrent.atomic.*
+import java.util.concurrent.locks.ReentrantLock
+import kotlin.coroutines.Continuation
+
+fun traverseTestObject(obj: Any): Map<Any, Int> {
+    val result = hashMapOf<Any, Int>()
+    traverseTestObject(obj, Collections.newSetFromMap(IdentityHashMap()), result)
+
+    return result
+}
+
+private fun traverseTestObject(obj: Any, visualized: MutableSet<Any>, result: MutableMap<Any, Int>) {
+    if (!visualized.add(obj)) return
+    val objectNumber = getObjectNumber(obj.javaClass, obj)
+    result[obj] = objectNumber
+
+    var clazz: Class<*>? = obj.javaClass
+    if (clazz!!.isArray) {
+        if (obj is Array<*>) {
+            obj.forEach { element -> element?.let { traverseTestObject(it, visualized, result) } }
+        }
+        return
+    }
+    while (clazz != null) {
+        clazz.declaredFields.filter { f ->
+            !Modifier.isStatic(f.modifiers) &&
+                    !f.isEnumConstant &&
+                    f.name != "serialVersionUID"
+        }.forEach { f ->
+            try {
+                try {
+                    f.isAccessible = true
+                } catch (_: Throwable) {
+                    return@forEach
+                }
+                var value: Any? = f.get(obj)
+
+                if (isAtomic(value)) {
+                    value = value!!.javaClass.getDeclaredMethod("get").invoke(value)
+                }
+
+                if (value?.javaClass?.canonicalName == "kotlinx.atomicfu.AtomicRef") value =
+                    value.javaClass.getDeclaredField("value").apply { isAccessible = true }.get(value)
+                if (value?.javaClass?.canonicalName == "kotlinx.atomicfu.AtomicInt") value =
+                    value.javaClass.getDeclaredField("value").apply { isAccessible = true }.get(value)
+                if (value?.javaClass?.canonicalName == "kotlinx.atomicfu.AtomicLong") value =
+                    value.javaClass.getDeclaredField("value").apply { isAccessible = true }.get(value)
+                if (value?.javaClass?.canonicalName == "kotlinx.atomicfu.AtomicBoolean") value =
+                    value.javaClass.getDeclaredField("_value").apply { isAccessible = true }.get(value)
```

Comment:
Please take a look how Lincheck reads fields in the `LincheckJavaAgent` class via `Unsafe`